src/test/spec/unified_runner/test_file.rs

```diff
@@ -0,0 +1,323 @@
+    pub run_on_requirements: Option<Vec<RunOnRequirement>>,
+    pub allow_multiple_mongoses: Option<bool>,
+    pub create_entities: Option<Vec<Entity>>,
+    pub initial_data: Option<Vec<CollectionData>>,
+    pub tests: Vec<TestCase>,
+}
+
+fn deserialize_schema_version<'de, D>(deserializer: D) -> std::result::Result<Version, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let mut schema_version = String::deserialize(deserializer)?;
+    // if the schema version only contains a major and minor version (e.g. 1.0), append a ".0" to
+    // ensure correct parsing into a semver::Version
+    if schema_version.split('.').count() == 2 {
+        schema_version.push_str(".0");
+    }
+    Version::parse(&schema_version).map_err(|e| serde::de::Error::custom(format!("{}", e)))
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct RunOnRequirement {
+    min_server_version: Option<String>,
+    max_server_version: Option<String>,
+    topology: Option<Vec<Topology>>,
+}
+
+#[derive(Debug, Deserialize, PartialEq)]
+#[serde(rename_all = "lowercase")]
+#[serde(untagged)]
+pub enum Topology {
+    Single,
+    ReplicaSet,
+    Sharded,
+    #[serde(rename = "sharded-replicaset")]
+    ShardedReplicaSet,
+}
+
+impl RunOnRequirement {
+    pub fn can_run_on(&self, client: &TestClient) -> bool {
+        if let Some(ref min_version) = self.min_server_version {
+            let req = VersionReq::parse(&format!(">= {}", &min_version)).unwrap();
+            if !req.matches(&client.server_version) {
+                return false;
+            }
+        }
+        if let Some(ref max_version) = self.max_server_version {
+            let req = VersionReq::parse(&format!("<= {}", &max_version)).unwrap();
+            if !req.matches(&client.server_version) {
+                return false;
+            }
+        }
+        if let Some(ref topology) = self.topology {
+            if !topology.contains(&client.topology()) {
+                return false;
+            }
+        }
+        true
+    }
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub enum Entity {
+    Client(Client),
+    Database(Database),
+    Collection(Collection),
+    Session(Session),
+    Bucket(Bucket),
+    Stream(Stream),
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Client {
+    pub id: String,
+    #[serde(
+        default = "default_uri",
+        deserialize_with = "deserialize_uri_options_to_uri_string",
+        rename = "uriOptions"
+    )]
+    pub uri: String,
+    pub use_multiple_mongoses: Option<bool>,
+    pub observe_events: Option<Vec<String>>,
+    pub ignore_command_monitoring_events: Option<Vec<String>>,
+}
+
+fn default_uri() -> String {
+    std::env::var("MONGODB_URI").unwrap_or_else(|_| "mongodb://localhost:27017/".to_string())
+}
+
+fn deserialize_uri_options_to_uri_string<'de, D>(
+    deserializer: D,
+) -> std::result::Result<String, D::Error>
+where
```

Comment:
`ClientOptions::parse` can't be used here since it's async, so instead the URI options are deserialized into a URI. This function takes care of merging the options in the existing URI with the test options (which was previously done when creating the client), so when the client gets created it now just needs to pass in this URI to `ClientOptions::parse`.